Anchor: Gutenboarding Launch works immediately without domain or plan selection

mreishus opened the pull request:
#### Changes proposed in this Pull Request

* After creating a site with the Anchor Gutenboarding process, the 'Launch' button displayed in the editor will launch the site without asking the user to select a domain or plan.

#### Note: Gutenboarding Launch Methods
* There are currently has 3 different ways to handle site launch. "Step By Step", "Focused Launch", and "Calypso Launch."
* The code that determines which launch style is used is [here](https://github.com/Automattic/wp-calypso/blob/5b0a61b0fc53680a98c2bdd7cb205fc245f7e496/apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts#L120). 
* "Focused Launch" looks to be a newer experiment that is only turned on with a feature flag, and there are references to future A/B tests in the comments.
* Currently, free sites made with gutenboarding are sent to "Step By Step" launch. This covers all anchor gutenboarding sites.
  * The code in this PR only covers "Step by Step" launch.
  * I added an extra conditional to always send Anchor Gutenboarding sites to "Step by Step" launch: c7608994698fe6. While this condition is currently superfluous, my goal is to make Anchor Gutenboarding more robust against future changes by other teams.  However, I'm open to removing it if it's considered unneeded complexity.

#### Testing instructions

* Apply patch D56138-code.
* Need to build and sync `apps/editing-toolkit` from this branch to sandbox
  * In a terminal tab:
  * `cd ./apps/editing-toolkit`
  * `yarn dev --sync`
  * Changes should be in `wp-content/plugins/full-site-editing-plugin/dev/editor-gutenboarding-launch/dist/editor-gutenboarding-launch.js` & also `launch-button.js`.
    * If changes are there and it's not working, edit a post, click the vertical three dots in the top right, and at the bottom of the sidebar, verify the version of the editing toolkit being loaded is `dev`
* Need to create new site using anchor gutenboarding
  * Run calypso locally
  * Sandbox public-api
  * Visit http://calypso.localhost:3000/new?&anchor_podcast=22b6608&anchor_episode=e324a06c-3148-43a4-85d8-34c0d8222138&spotify_url=https%3A%2F%2Fopen.spotify.com%2Fshow%2F6HTZdaDHjqXKDE4acYffoD%3Fsi%3DEVfDYETjQCu7pasVG5D73Q and create a new site. You can either use incognito and make a new account or log into an existing account.
  * Go through all anchor gutenboarding steps.
  * You should end on an editor page with the podcast player embedded.  Keep that tab/URL.
* Need to sandbox that exact site you just created
  * Check the URL to see the site you created, and add that to your /etc/hosts to point to your sandbox.  For example, it might look like `bravenotperfectwithreshmasaujani340284555.wp.com`
  * Clear DNS cache in browser OR restart browser OR wait 3-5 minutes
    * Firefox: Visit `about:networking#dns` -> Clear Cache
    * You might have to close, reopen and relogin with any incognito windows? I think they keep their own DNS cache that can't be cleared for some reason.
  * Visiting `/wp-admin` on it should show sandboxed banner
  * F5 your editor page. Very important: If you press the three vertical dots in the top right, the bottom of the sidebar should read `Editing Toolkit: dev`. If it doesn't, you won't see the changes here.
    * I have struggled to get this working even though I have done everything "right" in setup.  You can try restarting the browser, or restarting autoproxxy, or as a hacky workaround, copy the contents of `wp-content/plugins/full-site-editing-plugin/dev/` in your sandbox over `wp-content/plugins/full-site-editing-plugin/49211111111/` (change number to the one being displayed).
* Press the Launch button in the top right
  * Your site should launch immediately without being asked for domain or plan upsells
* Test the negative: (Does this PR only affect anchor?) 
  * Create a new gutenboarding site that isn't anchor (Use only http://calypso.localhost:3000/new )
  * Go through gutenboarding steps
  * Sandbox that new site
  * Land on editor page, save url
  * Ensure the dev version of the FSE plugin is loading
  * On the editor page, Click launch button
  * Now you should see the existing domain and plan upsells (they should be hidden on the anchor path only)

fixes 391-gh-Automattic/dotcom-manage

mreishus committed:
WIP

mreishus commented:
This is blocked on 416-gh-Automattic/dotcom-manage
 Edit: Resuming work now that D55472-code is available

mreishus committed:
Site option name anchor_fm_podcast_id -> anchor_podcast

mreishus committed:
Add isLaunchImmediately to LaunchModal

This needs to be behind an anchor gutenboarding check, not always
running.

mreishus committed:
Only launchImmediately for anchor gutenboarding

mreishus commented:
This is currently working for me, with one bug: When I click launch, it asks me if I want to navigate away.  Possibly there's not an autosave going off somewhere.
![2021-01-18_12-15](https://user-images.githubusercontent.com/937354/104951094-84e52980-5987-11eb-9f5c-954c0ae1b0e4.png)

Because of the setTimeout, the order is
1. Open Sidebar https://github.com/Automattic/wp-calypso/blob/a48595c6aabe4083a32f7d109fabfb0d35a2debb/apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts#L124
2. Start Launching https://github.com/Automattic/wp-calypso/blob/a48595c6aabe4083a32f7d109fabfb0d35a2debb/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-modal/index.tsx#L44
3. (Are you sure you want to leave? Modal Displays)
4. Save post https://github.com/Automattic/wp-calypso/blob/a48595c6aabe4083a32f7d109fabfb0d35a2debb/apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts#L126

mreishus committed:
handleLaunch() is delayed 2 seconds to allow post to save

mreishus commented:
* ~Is basically ready for review, but I would prefer if we had a solution to #416 first.~
  *  ~Waiting for  Automattic/wp-calypso#49018 to be merged~
  * ~Waiting for https://github.com/Automattic/wp-calypso/pull/48901 to be merged~
  * ~Waiting for D55324-code to be merged~
* Edit: The above is fixed and this is now unblocked.  I will be reviewing it without my sandbox hacks and with prod code. 
* ~~Debugging code is still included but that may make it easier to review, since it can be tricky.~~ Removed debugging in c13f89515ab9d, temporarily revert this to add some

mreishus committed:
Add savePost to dependency array

mreishus committed:
Remove debugging

krymson24 commented:
I tried a couple of things to try and get this to work -- in case I was misunderstanding the instructions, but I think it may not be working for me because of the same struggles you were experiencing, or the script that you wrote. I get this error:

```
sent 29 bytes  received 20 bytes  32.67 bytes/sec
total size is 0  speedup is 0.00
rsync error: some files could not be transferred (code 23) at /AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Sources/rsync/rsync-55/rsync/main.c(996) [sender=2.6.9]
```

I also replaced `inotifywait` for `fswatch` here `while fswatch -r "$LOCAL_PATH"; do`

On an existing post `https://blazblazpodcast.wordpress.com/wp-admin/post.php?post=3&action=edit`, I see
<img width="1666" alt="Screen Shot 2021-01-20 at 10 13 52 PM" src="https://user-images.githubusercontent.com/66652282/105289379-1cb06680-5b86-11eb-9e0d-7c3276ae1acb.png">

For a new post
![2021-01-20 22 08 21](https://user-images.githubusercontent.com/66652282/105289762-35b91780-5b86-11eb-8052-677c686223b0.gif)

This is on `https://wordpress.com/post/blazblazpodcast.wordpress.com/3`
![2021-01-20 22 11 39](https://user-images.githubusercontent.com/66652282/105290236-56816d00-5b86-11eb-988f-d90c420983d9.gif)

Any suggestions on how I can test this correctly? I can come back to this tomorrow!

krymson24 commented:
Also love your persistence on this @mreishus! I probably would have gotten extremely frustrated with the set-up. Awesome work! 👍

krymson24 commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-modal/index.tsx:
```diff
@@ -32,11 +33,25 @@ const LaunchModal: React.FunctionComponent< Props > = ( { onClose } ) => {
+			setIsLaunching( true );
+			savePost();
+
+			// Delay handleLaunch() to allow for the save to complete;
```
😞

blackjackkent replied:
This seems...less than ideal. I'm not super familiar with @wordpress/data stuff, but is there some sort of promise-based way to await the result of saving the post?

https://developer.wordpress.org/block-editor/packages/packages-data/ I haven't looked into it in depth but the `subscriber` section of this documentation seems like it could be promising.

simison replied:
See example elsewhere:

https://github.com/Automattic/wp-calypso/blob/61acc957638fc7009632af4c0c058aa6bc0e82b7/apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts#L79

Is that not working?

mreishus replied:
> See example elsewhere:
> 
> https://github.com/Automattic/wp-calypso/blob/61acc957638fc7009632af4c0c058aa6bc0e82b7/apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts#L79
> 
> Is that not working?

Hey thanks, that worked. Done here: 93cf373d36ba. I believe I tried something like this early on, but I was scared away from a warning displayed in my editor:  `80007: 'await' has no effect on the type of this expression.`. It shows that when I look at the example linked above, too (`wp-calypso/apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts`, Line 79 in `61acc95`)

However, given that
* I can't get `eslint` or `tsc` to show those warnings on the command line
* It works in my testing

I'm going to consider it to be either a misconfigured editor or over-enthusiastic warning.  I see the warning in both vscode and vim + vim-coc.

kwight commented on apps/editing-toolkit/editing-toolkit-plugin/editor-gutenboarding-launch/index.ts:
```diff
@@ -94,8 +96,16 @@ function updateEditor() {
 			 */
-			const shouldOpenStepByStepLaunch = isGutenboarding && getIsFreePlan();
+			const shouldOpenStepByStepLaunch =
+				( isGutenboarding && getIsFreePlan() ) || isGutenboardingAnchorFm;
```
Great, I like this idea.

mreishus commented:
RE: Henry, sorry I couldn't respond in time, I shifted my efforts to an urgent problem elsewhere.  I'm now able to look at this again and Henry's working on a different project. :)

mreishus commented:
I will be investigating an alternative to setTimeout() shortly (subscriber in wp/data, if that doesn't work I am out of ideas).

In the meantime, after rebasing I'm finding the original workflow doesn't work anymore. I can create my new site, but after I sandbox it, close my incognito window, and re-open this url in a fresh incognito with no state:

```
http://calypso.localhost:3000/post/bravenotperfectwithreshmasaujaniXXXXXX.word press.com?anchor_podcast=22b6608&anchor_episode=e324a06c-3148-43a4-85d8-34c0d8222138&spotify_show_url=https://open.spotify.com/show/6HTZdaDHjqXKDE4acYffoD?si=EVfDYETjQCu7pasVG5D73Q
(Remove space between "word" and "press")
```

I would expect to see a log in page that redirects me back to my editor.  Instead, I see a blank white screen with lots of console errors. 

I believe the relevant one is here: 
```
Uncaught TypeError: URL constructor: /post/bravenotperfectwithreshmasaujaniXXXXXX.word press.com?anchor_podcast=22b6608&anchor_episode=e324a06c-3148-43a4-85d8-34c0d8222138&spotify_show_url=https://open.spotify.com/show/6HTZdaDHjqXKDE4acYffoD?si=EVfDYETjQCu7pasVG5D73Q is not a valid URL.
    getIsAnchorFmSignup app:///./landing/gutenboarding/utils.ts:10
(Remove space between "word" and "press")
```

Will be looking into this. This fix may go into a separate PR if I determine it's not related to these changes.
Created issue: 442-gh-Automattic/dotcom-manage

mreishus commented:
Blocked on 442-gh-Automattic/dotcom-manage

kwight commented:
That space is definitely a problem – are you saying that space was found in a generated URL in committed code somewhere? Is it in your config for the Chrome extension maybe?..

mreishus commented:
Ah, no, the space is just something I added to my comment to stop notifications. The actual urls generated have no space in them, and the error is still happening.

kwight commented:
Hm, I went through the logic of `getIsAnchorFmSignup` with your URL, and didn't hit any snags. 🤔 The error shows a relative URL, I don't know if that's part of it (I tested with the full URL). Does something strip the URL base?

mreishus commented:
I believe it only happens when you're logged out, that might be it? If I open a new incognito window with no state, and visit:

http://calypso.localhost:3000/post/bravenotperfectwithreshmasaujaniXXXXXX.sentencepress.com?anchor_podcast=22b6608&anchor_episode=e324a06c-3148-43a4-85d8-34c0d8222138&spotify_show_url=https://open.spotify.com/show/6HTZdaDHjqXKDE4acYffoD?si=EVfDYETjQCu7pasVG5D73Q

I'm redirected to

http://calypso.localhost:3000/log-in?site=bravenotperfectwithreshmasaujaniXXXXXX.sentencepress.com&redirect_to=%2Fpost%2FbravenotperfectwithreshmasaujaniXXXXXX.sentencepress.com%3Fanchor_podcast%3D22b6608%26anchor_episode%3De324a06c-3148-43a4-85d8-34c0d8222138%26spotify_show_url%3Dhttps%3A%2F%2Fopen.spotify.com%2Fshow%2F6HTZdaDHjqXKDE4acYffoD%3Fsi%3DEVfDYETjQCu7pasVG5D73Q

Which provides the error.  I'm about to post a PR for it 
Edit: https://github.com/Automattic/wp-calypso/pull/49280 -> Deployed

mreishus commented:
Unblocked

yansern commented:
Hello there! I'd like to mention that with https://github.com/Automattic/wp-calypso/pull/49004, D55512-code and also https://github.com/Automattic/wp-calypso/pull/49067 we are actually moving away from using the client-side approach of determining which Launch version to use and moving all that logic to the server-side. We find this approach cleaner and much simplier to maintain/deploy in the long run.

A quick glance on the changes in this PR, the way an AnchorFM WP site is determined using `getSiteOption( state, siteId, 'anchor_podcast' )`, since site option can be retrieved in PHP, it is actually possible to do so on the PHP end.

We'll need to coordinate how to best move forward. If our PRs gets merged first I can help with porting the code changes I see here over to the new approach.

simison commented:
@yansern  Considering focused launch is almost ready to ship, and Anchor is a month away, I'd rather get focused launch approach in first and then refactor here.

razvanpapadopol commented:
> we are actually moving away from using the client-side approach of determining which Launch version to use and moving all that logic to the server-side

This is also allowing us to enable the Launch button and editor Launch flows in wp-admin 🎉

mreishus committed:
Use await savePost to remove setTimeout() hack

mreishus commented:
This is basically ready to go, except for a review OK on one small change: 93cf373.  However, it sounds like it needs to be reworked completely after the focus launch changes hit. Do we know when those will be merged in, and if they're still a WIP, can we at least add the isGutenboardingAnchorFm flag right now?

yansern committed:
Move over to new launch button server-side style.

yansern committed:
Remove isGutenboardingAnchorFm from iframe bridge server.

yansern commented:
I've rebased and updated this PR with the following changes:

1. The `anchorFmPodcastId` is first retrieved in D56138-code on the server-side through the `a8c_enable_editor_site_launch` filter.

2. The `editor-site-launch/index.php` calls the `a8c_enable_editor_site_launch`  filter, retrieves the `anchorFmPodcastId` and injects it into the `window.wpcomEditorSiteLaunch` object via `wp_localize_script`. See https://github.com/Automattic/wp-calypso/pull/48938/files#diff-ed674ad5d1a1fd12d531950e48e3b988ce0cb0593e18a24a206b1da82cca564eR57-R61.

3. The launch button script (which is now moved to `editor-site-launch/src/launch-button/index.ts`) retrieves the `anchorFmPodcastId` via the aforementioned `window.wpcomEditorSiteLaunch` object. 

4. If the `anchorFmPodcastId` exists, it will then `dispatch( 'automattic/launch' ).enableAnchorFm();`. See https://github.com/Automattic/wp-calypso/pull/48938/files#diff-5ca2166fe3b2f3927d78bb85557a2582921420cc4838e80c276670d324ee504dR60-R65

Also, `isGutenboardingAnchorFm` is removed from iframe bridge server, so this PR no longer need to build & sync `apps\wpcom-block-editor` (yay, one less thing to do in this complicated setup!).

I'll do some light updating on the PR description above as well.

yansern commented:
As I was updating the PR description above, I'd like to add following thoughts.

> * Need to build and sync `apps/wpcom-block-editor` from this branch to sandbox
>  * ~~`yarn dev --sync`~~ <-- Doesn't work for me, so I used two terminals here
>  * terminal 1: `cd ./apps/wpcom-block-editor`, `yarn dev`
>  * terminal 2: `cd ./apps/wpcom-block-editor`, `for i in {1..10}; do bin/wpcom-rsync.sh; done` (Custom rsync script, text ?here: 27ee8-pb , this works for me while the existing chokidar one does not)'
>    * loop is to restart it when it crashes (when dist/ directory is deleted and rebuilt on change) 

I removed the above step as compiling `wpcom-block-editor` is no longer required.

As for why running `yarn dev --sync` doesn't sync to your sandbox, check if you have both the `wpcom-sandbox` and `wpcom` host alias pointing to your sandbox in `~/.ssh/config`. Yes, both alias. If I'm not mistaken, ETK is using one, and WBE is using the other.

> I have struggled to get this working even though I have done everything "right" in setup. You can try restarting the browser, or restarting autoproxxy, or as a hacky workaround, copy the contents of wp-content/plugins/full-site-editing-plugin/dev/ in your sandbox over wp-content/plugins/full-site-editing-plugin/49211111111/ (change number to the one being displayed).

As for this one, the **AutoProxxy > Force Reconnection** will work, but only if you do not have any other forms of proxy setup in place. I used to use the FalconProxy chrome extension to proxy into A8C. When I did that, browser's DNS cache clearing did not work. I always had to close and reopen browsers. Since I removed FalconProxy and switch over to solely using AutoProxxy, I do not have to go through all that hassle anymore.

mreishus commented:
Thanks for the update! 
Currently, my `yarn dev --sync` is crashing inside `./apps/editing-toolkit`.

(Logging info removed, check history)

**Edit: Fixed with this hack.**
```
cd vendor/bin
mv phpcs phpcs.old
mv phpcbf phpcbf.old
ln -s ../squizlabs/php_codesniffer/bin/phpcbf phpcbf
ln -s ../squizlabs/php_codesniffer/bin/phpcs phpcs
```

:shrug:

mreishus commented:
This PR + D56138-code is now correctly skipping the plan and domain steps while launching, however the launch doesn't work anymore.

Click launch -> See interstitial (same as before)
![2021-01-28_10-45](https://user-images.githubusercontent.com/937354/106170445-14dc5d80-6156-11eb-9195-360ea55426ee.png)

Interstitial now goes to a failing checkout page
![2021-01-28_10-45_1](https://user-images.githubusercontent.com/937354/106170496-245ba680-6156-11eb-90f9-524fc2af6a12.png)

yansern commented:
> This PR + D56138-code is now correctly skipping the plan and domain steps while launching, however the launch doesn't work anymore.

I have a feeling since Domains & Plans step was skipped.. no plans was selected so no plan was added to the cart.

What plan is an anchor site on by default on WordPress?

kwight commented:
> What plan is an anchor site on by default on WordPress?

All Anchor sites are to be launched with only a free plan, per our agreement with them (and we're not to try to upgrade them during the process).

mreishus commented:
I'm finding that the plan selected is undefined, which is not the free plan, so some javascript code sends the user to checkout.  I would like the plan selected to be the free plan.  I'm trying to do it in this diff, but it doesn't work:

```diff
diff --git a/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-modal/index.tsx b/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-modal/index.tsx
index e927c45dcb..11cb51e3d2 100644
--- a/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-modal/index.tsx
+++ b/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-modal/index.tsx
@@ -8,11 +8,12 @@ import { __ } from '@wordpress/i18n';
 import { Modal, Button } from '@wordpress/components';
 import { Icon, wordpress, close } from '@wordpress/icons';
 import { LaunchContext, useOnLaunch } from '@automattic/launch';
+import { useLocale } from '@automattic/i18n-utils';
 
 /**
  * Internal dependencies
  */
-import { LAUNCH_STORE, SITE_STORE } from '../stores';
+import { LAUNCH_STORE, SITE_STORE, PLANS_STORE } from '../stores';
 import Launch from '../launch';
 import LaunchSidebar from '../launch-sidebar';
 import LaunchProgress from '../launch-progress';
@@ -35,7 +36,14 @@ const LaunchModal: React.FunctionComponent< Props > = ( { onClose, isLaunchImmed
 	const { launchSite } = useDispatch( SITE_STORE );
 	const { savePost } = useDispatch( 'core/editor' );
 
+	const locale = useLocale();
+	const { setPlanProductId } = useDispatch( LAUNCH_STORE );
+	const defaultFreePlan = useSelect( ( select ) =>
+		select( PLANS_STORE ).getDefaultFreePlan( locale )
+	);
+
 	const handleLaunch = React.useCallback( () => {
+		console.log( 'handleLaunch, launching..' );
 		launchSite( siteId );
 		setIsLaunching( true );
 	}, [ launchSite, setIsLaunching, siteId ] );
@@ -45,9 +53,12 @@ const LaunchModal: React.FunctionComponent< Props > = ( { onClose, isLaunchImmed
 			await savePost();
 		};
 		if ( ! isLaunching && isLaunchImmediately ) {
-			setIsLaunching( true );
-			asyncSavePost();
-			handleLaunch();
+			console.log( 'anchor hook' );
+			setPlanProductId( defaultFreePlan );
+			console.log( 'anchor hook set plan 2', { defaultFreePlan } );
+			/* setIsLaunching( true ); */
+			/* asyncSavePost(); */
+			/* handleLaunch(); */
 		}
 	}, [ isLaunching, isLaunchImmediately, handleLaunch, savePost ] );
 
diff --git a/packages/data-stores/src/site/actions.ts b/packages/data-stores/src/site/actions.ts
index 061bc71165..3b8a378b7e 100644
--- a/packages/data-stores/src/site/actions.ts
+++ b/packages/data-stores/src/site/actions.ts
@@ -116,11 +116,13 @@ export function createActions( clientCreds: WpcomClientCredentials ) {
 	function* launchSite( siteId: number ) {
 		yield launchSiteStart( siteId );
 		try {
+			console.log('Making wpcom request to launch endpoint');
 			yield wpcomRequest( {
 				path: `/sites/${ siteId }/launch`,
 				apiVersion: '1.1',
 				method: 'post',
 			} );
+			console.log('Yielding launch Site Success');
 			yield launchSiteSuccess( siteId );
 		} catch ( _ ) {
 			yield launchSiteFailure( siteId, SiteLaunchError.INTERNAL );
diff --git a/packages/launch/src/hooks/use-on-launch.ts b/packages/launch/src/hooks/use-on-launch.ts
index f01fc96779..3d4edc6366 100644
--- a/packages/launch/src/hooks/use-on-launch.ts
+++ b/packages/launch/src/hooks/use-on-launch.ts
@@ -24,15 +24,21 @@ export const useOnLaunch = () => {
 	const { isSiteLaunched } = useSite();
 	const { siteSubdomain } = useSiteDomains();
 	const { goToCheckout } = useCart();
+	console.log( 'useOnLaunch: ', { isSiteLaunched, planProductId, isPlanFree } );
 
 	React.useEffect( () => {
+		console.log( 'useOnLaunch: hook: isSiteLaunched changed' );
+		console.log( { isSiteLaunched, isPlanFree } );
 		if ( isSiteLaunched ) {
 			// if a paid plan is selected, set cart and redirect to /checkout
 			if ( ! isPlanFree ) {
+				console.log( 'Going to checkout' );
 				goToCheckout();
 				return;
 			}
 			// if free plan is selected, redirect to My Home
+			console.log( 'Going here:' );
+			console.log( `/home/${ siteSubdomain?.domain || siteId }` );
 			redirectTo( `/home/${ siteSubdomain?.domain || siteId }` );
 		}
 	}, [ isSiteLaunched ] ); // eslint-disable-line react-hooks/exhaustive-deps
```

I've added some debugging and modified the anchor "launch immediately" hook to not actually launch, but to try to set the free plan. However, my value of `defaultFreePlan` ends up being undefined here. Maybe I need to request all of the plans be loaded first?

mreishus commented:
Ok, think I have a possible solution. Above, I am setting the plan*ProductId* to an object, not a number. And some dependency problems that I am solving.

mreishus committed:
Anchor Gutenboarding Launch: Skip checkout page

- Set the plan available to the free plan, so the user is not sent to
  the checkout page.
- Problem: User sees the domain upsell page for a few moments while the
  plans are still loading.

mreishus committed:
Anchor Gutenboarding Launch: Solve "Flash of Domain Upsell" problem

mreishus commented:
## Ready for Review again

I've gotten it working again after the rebase and the underlying changes to launch. I had to make these two changes:
* 1cc379bfa7 Anchor Gutenboarding Launch: Skip checkout page
* f22b5a8d48 Anchor Gutenboarding Launch: Solve "Flash of Domain Upsell" problem 

This does increase the complexity of `apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-modal/index.tsx` by adding another useEffect and another state variable, but I don't see another way to solve it.  Any alternative approaches are welcomed.

### Assumption Made
* The free plan will only have one product id and it is safe to `setPlanProductId( defaultFreePlan?.productIds[ 0 ] );`. Is there a better way to set the plan to the free plan?

kwight reviewed (approved):
Fantastic – everything worked as expected first time, both Anchor and regular signup flows. 🎉
